x-pack/filebeat/input/gcs/metrics.go

```diff
@@ -0,0 +1,72 @@
+
+package gcs
+
+import (
+	"github.com/rcrowley/go-metrics"
+
+	"github.com/elastic/beats/v7/libbeat/monitoring/inputmon"
+	"github.com/elastic/elastic-agent-libs/monitoring"
+	"github.com/elastic/elastic-agent-libs/monitoring/adapter"
+)
+
+// inputMetrics handles the input's metric reporting.
+type inputMetrics struct {
+	unregister        func()
+	url               *monitoring.String // URL of the input resource
+	errorsTotal       *monitoring.Uint   // number of errors encountered
+	decodeErrorsTotal *monitoring.Uint   // number of decode errors encountered
+
+	gcsObjectsRequestedTotal        *monitoring.Uint // Number of GCS objects downloaded.
+	gcsObjectsPublishedTotal        *monitoring.Uint // Number of GCS objects processed that were published.
+	gcsObjectsListedTotal           *monitoring.Uint // Number of GCS objects returned by list operations.
+	gcsBytesProcessedTotal          *monitoring.Uint // Number of GCS bytes processed.
+	gcsEventsCreatedTotal           *monitoring.Uint // Number of events created from processing GCS data.
+	gcsFailedJobsTotal              *monitoring.Uint // Number of failed jobs.
+	gcsExpiredFailedJobsTotal       *monitoring.Uint // Number of expired failed jobs that could not be recovered.
+	gcsObjectsInflight              *monitoring.Uint // Number of GCS objects inflight (gauge).
+	gcsObjectProcessingTime         metrics.Sample   // Histogram of the elapsed GCS object processing times in nanoseconds (start of download to completion of parsing).
+	gcsObjectSizeInBytes            metrics.Sample   // Histogram of processed GCS object size in bytes
+	gcsEventsPerObject              metrics.Sample   // Histogram of events in an individual GCS object
+	gcsJobsScheduledAfterValidation metrics.Sample   // Number of jobs scheduled after validation.
+}
+
+func newInputMetrics(id string, optionalParent *monitoring.Registry) *inputMetrics {
+	reg, unreg := inputmon.NewInputRegistry(inputName, id, optionalParent)
+	out := &inputMetrics{
+		unregister:        unreg,
+		url:               monitoring.NewString(reg, "url"),
+		errorsTotal:       monitoring.NewUint(reg, "errors_total"),
+		decodeErrorsTotal: monitoring.NewUint(reg, "decode_errors_total"),
+
+		gcsObjectsRequestedTotal:        monitoring.NewUint(reg, "gcs_objects_requested_total"),
+		gcsObjectsPublishedTotal:        monitoring.NewUint(reg, "gcs_objects_published_total"),
+		gcsObjectsListedTotal:           monitoring.NewUint(reg, "gcs_objects_listed_total"),
+		gcsBytesProcessedTotal:          monitoring.NewUint(reg, "gcs_bytes_processed_total"),
+		gcsEventsCreatedTotal:           monitoring.NewUint(reg, "gcs_events_created_total"),
+		gcsFailedJobsTotal:              monitoring.NewUint(reg, "gcs_failed_jobs_total"),
+		gcsExpiredFailedJobsTotal:       monitoring.NewUint(reg, "gcs_expired_failed_jobs_total"),
+		gcsObjectsInflight:              monitoring.NewUint(reg, "gcs_objects_inflight"),
```

Comment:
Suffix this with `_gauge` to make it clear that it should be treated as a gauge.